packages/core/src/tracing/instrumentation/database/db2.js

```diff
@@ -0,0 +1,539 @@
+      const originalFn = result[fnName];
+
+      result[fnName] = function instanaSyncOverride() {
+        try {
+          const res = originalFn.apply(this, arguments);
+
+          if (res instanceof Error) {
+            span.ec = 1;
+            span.data.db2.error = tracingUtil.getErrorDetails(res);
+          }
+          return res;
+        } catch (err) {
+          span.ec = 1;
+          span.data.db2.error = tracingUtil.getErrorDetails(err);
+
+          return err;
+        }
+      };
+    }
+  });
+}
+
+function instrumentQueryHelper(ctx, originalArgs, originalFunction, stmt, isAsync) {
+  if (skipTracing()) {
+    return originalFunction.apply(ctx, originalArgs);
+  }
+
+  const parentSpan = cls.getCurrentSpan();
+
+  // CASE: There can be only one exit span, skip
+  if (constants.isExitSpan(parentSpan)) {
+    return originalFunction.apply(ctx, originalArgs);
+  }
+
+  return cls.ns.runAndReturn(() => {
+    const span = createSpan(stmt, instrumentQueryHelper);
+
+    // CASE: querySync
+    if (!isAsync) {
+      try {
+        const result = originalFunction.apply(ctx, originalArgs);
+
+        if (result instanceof Error) {
+          span.ec = 1;
+          span.data.db2.error = tracingUtil.getErrorDetails(result);
+        }
+
+        finishSpan(ctx, result, span);
+        return result;
+      } catch (e) {
+        span.ec = 1;
+        span.data.db2.error = tracingUtil.getErrorDetails(e);
+        finishSpan(ctx, null, span);
+        throw e;
+      }
+    }
+
+    const customerCallbackIndex =
+      // eslint-disable-next-line no-nested-ternary
+      originalArgs.length === 2 && typeof originalArgs[1] === 'function'
+        ? 1
+        : originalArgs.length === 3 && typeof originalArgs[2] === 'function'
+        ? 2
+        : null;
+    const customerCallback = originalArgs[customerCallbackIndex];
+
+    if (customerCallback) {
+      originalArgs[customerCallbackIndex] = function instanaCallback(err) {
+        if (err) {
+          span.ec = 1;
+          span.data.db2.error = tracingUtil.getErrorDetails(err);
+        }
+
+        finishSpan(ctx, null, span);
+        return customerCallback.apply(this, arguments);
+      };
+
+      return originalFunction.apply(ctx, originalArgs);
+    }
+
+    const resultPromise = originalFunction.apply(ctx, originalArgs);
+
+    resultPromise
+      .then(result => {
+        finishSpan(ctx, result, span);
+        return result;
+      })
+      .catch(err => {
+        span.ec = 1;
+        span.data.db2.error = tracingUtil.getErrorDetails(err);
+        finishSpan(ctx, null, span);
+        return err;
+      });
+
+    return resultPromise;
+  });
+}
+
+function instrumentExecuteHelper(ctx, originalArgs, stmtObject, httpParentSpan) {
+  const originalExecuteNonQuerySync = stmtObject.executeNonQuerySync;
+
+  stmtObject.executeNonQuerySync = function instanaExecuteNonQuerySync() {
+    return cls.ns.runAndReturn(() => {
+      cls.setCurrentSpan(httpParentSpan);
```

Comment:
_question:_ Since we cannot assume that the statement is prepared in the context of any HTTP request, we might need to rethink when and how we capture the CLS context/parent span and how we can restore it. Let's discuss that face to face.

Reply:
I guess we can take the parenSpan from the execute call if that is doable 🤔

Reply:
Done 👍